extensions/parquet/compression/src/main/java/io/deephaven/parquet/compress/LZ4WithLZ4RawBackupCompressorAdapter.java

```diff
@@ -51,17 +55,23 @@ public BytesInput decompress(final InputStream inputStream, final int compressed
         // Buffer input data in case we need to retry with LZ4_RAW.
         final BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream, compressedSize);
         bufferedInputStream.mark(compressedSize);
-        BytesInput ret;
         try {
-            ret = super.decompress(bufferedInputStream, compressedSize, uncompressedSize, decompressorCache);
+            // Try to decompress the bytes with LZ4 first.
+            final InputStream decompressedInput =
+                    super.decompress(bufferedInputStream, compressedSize, uncompressedSize, decompressorCache);
+            final ByteBuffer decompressedBuffer =
+                    CompressorAdapter.readNBytes(decompressedInput, uncompressedSize, new byte[uncompressedSize]);
```

Comment:
It seems very sad that we need to do the full decompression to figure out whether it's LZ4 or LZ4_RAW. 

We should either handle the LZ4/LZ4_RAW exception handling at a higher layer (so we don't need to materialize all at once into ByteBuffer), or we should have a specialized InputStream that could do the reset() + fallback internally without the all-at-once read. Is there a hard limit on how many bytes it takes to fail an LZ4_RAW mislabelled as LZ4?

Reply:
I don't think there is an exact limit that I can check here, so just checking for a failure.
And yea, I couldn't find an easy way to do this without making a bigger change, and like Ryan said, this extra double buffering will happen only once.

Reply:
Maybe there is room to have a better impl (unified InputStream) return in the future. Or, some way for the user to disable this fallback logic (so if they trust their parquet when it says LZ4, they don't have to pay this extra buffering cost).